src/wasi.ts

```diff
@@ -38,6 +40,19 @@ export default class WASI {
       } else {
         throw e;
       }
+    } finally {
+      // Close all fds except stdin/stdout/stderr and the preopen
+      // directories. No need to call fd_close on individual fds, just
+      // reset the fds/freeFds array, and drop all the non-preopen fds
+      // that follow the preopen ones:
+      // https://github.com/WebAssembly/wasi-libc/blob/wasi-sdk-27/libc-bottom-half/sources/preopens.c#L244.
+      this.#freeFds = [];
+      for (let fd = 3; fd < this.fds.length; ++fd) {
+        if (!(this.fds[fd] instanceof PreopenDirectory)) {
+          this.fds.splice(fd, this.fds.length - fd);
+          break;
+        }
+      }
```

Comment:
A single `WASI` instance shouldn't be reused across multiple `.start()` calls, so this can close all fds. This should still call `fd_close` on individual fds though, they may want a custom cleanup.

I'm a bit concerned about breaking existing users that may be reusing `WASI` across multiple `.start()` calls though. Maybe moving this commit to a separate PR that could be merged for a future v0.3 release would be a solution?

Reply:
Fair. I've removed the commit and will open a follow-up one after this one lands.